src/nvim/winfloat.c

```diff
@@ -291,12 +296,138 @@ bool win_float_valid(const win_T *win)
+  api_free_object(obj);
+  xfree(data);
+}
+
+static void parse_title_pos(FloatConfig *fconfig, BorderTextType bt, char *s, size_t len)
+{
+  int start = bt == kBorderTextTitle ? 10 : 11;
+  Error err = ERROR_INIT;
+  String str = {
+    .data = xmemdupz(s + start, len),
+    .size = len + 1,
+  };
+  parse_bordertext_pos(str, bt, fconfig, &err);
+  api_free_string(str);
+}
+
+/// used for parse completepopup and previewpopup options
+int parse_float_option(FloatConfig *fconfig, int *border_adj, bool preview)
+{
+  char *p = p_pvp;
+  Error err = ERROR_INIT;
+
+  for (; *p != NUL; p += (*p == ',' ? 1 : 0)) {
+    char *e, *dig;
+    char *s = p;
+
+    e = vim_strchr(p, ':');
+    if (e == NULL || e[1] == NUL) {
+      return FAIL;
+    }
+
+    p = vim_strchr(e, ',');
+    if (p == NULL) {
+      p = e + strlen(e);
+    }
+    dig = e + 1;
+
+    if (strncmp(s, "border:", 7) == 0) {
+      if ((s + 7)[0] != 'd' && (s + 7)[0] != 's' && (s + 7)[0] != 'r') {
+        return FAIL;
+      }
+      Object style = OBJECT_INIT;
+      size_t len = p ? (size_t)(p - s - 7) : (size_t)(s - 7);
+      char *data = xmemdupz(s + 7, len);
+      style.type = kObjectTypeString;
+      style.data.string.data = xstrdup(data);
+      style.data.string.size = strlen(data);
+      parse_border_style(style, fconfig, &err);
+      int border_attr = syn_name2attr("FloatBorder");
+      for (int i = 0; i < 8; i++) {
+        fconfig->border_attr[i] = fconfig->border_hl_ids[i]
+                                  ? hl_get_ui_attr(0, HLF_BORDER, fconfig->border_hl_ids[i], false)
+                                  : border_attr;
+        if (i < 4 && fconfig->border_chars[2 * i + 1][0]) {
+          border_adj[i] = 1;
+        }
+      }
+      api_free_object(style);
+      xfree(data);
+    } else if (strncmp(s, "title:", 6) == 0) {
+      size_t len = p ? (size_t)(p - s - 6) : (size_t)(s - 6);
+      parse_title(fconfig, kBorderTextTitle, s + 6, len);
+    } else if (strncmp(s, "title_pos:", 10) == 0) {
+      size_t len = p ? (size_t)(p - s - 10) : (size_t)(s - 10);
+      parse_title_pos(fconfig, kBorderTextTitle, s, len);
+    } else if (strncmp(s, "footer:", 7) == 0) {
+      size_t len = p ? (size_t)(p - s - 7) : (size_t)(s - 7);
+      parse_title(fconfig, kBorderTextFooter, s + 7, len);
+    } else if (strncmp(s, "footer_pos:", 11) == 0) {
+      size_t len = p ? (size_t)(p - s - 11) : (size_t)(s - 11);
+      parse_title_pos(fconfig, kBorderTextFooter, s, len);
+    } else if (preview && strncmp(s, "height:", 7) == 0) {
+      fconfig->height = getdigits_int(&dig, false, 0);
+    } else if (preview && strncmp(s, "width:", 6) == 0) {
+      fconfig->width = getdigits_int(&dig, false, 0);
+    } else if (preview && strncmp(s, "style:", 6) == 0) {
+      if ((s + 6)[0] != 'm') {
+        return FAIL;
+      }
+      fconfig->style = kWinStyleMinimal;
+    } else {
+      return FAIL;
+    }
+  }
+
+  return OK;
+}
+
+win_T *win_float_find_preview(bool find_info)
```

Comment:
argument in here is wired. it's better make w_float_is_info to be receive flag. then use flag here. can refactor on https://github.com/neovim/neovim/pull/26739 .